### ISSUE TITLE
PHP 7

### ISSUE THREAD

**Issue by kepford** [opendevshop/devshop#188]
Is it possible to use PHP version 7? If so, how would I go about changing to that?

**Comment by arttus** [comment]
It's really more a case of Drupal 7 supporting PHP 7. https://www.drupal.org/node/2656548. 

I'm not certain if there are Aegir or Devshop modules which depend on some older PHP 5 API's. Aegir has been around for many years so I wouldn't be surprised if there was.

What specific need are you trying to meet with PHP7?

**Comment by kepford** [comment]
The one of the dependencies in the Drupal 8 site I'm working on requires 7. But in general the PHP 5 life cycle is coming to a close and I think there is a big need to be able to use PHP 7. One can run PHP 7 on the popular Drupal platform providers as well. That said, maybe a better option would be to allow different versions of PHP to be used on the Aegir server as BOA Aegir does.

**Comment by arttus** [comment]
Yeah good point, sounds like what we really need is the ability to set which version of PHP per project.

**Comment by madalinignisca** [comment]
I think this needs to be solved in the core of Aegir project as it isn't supporting multiple php versions yet.

**Comment by arttus** [comment]
The AEgir specific issue is which version of php will AEgir run on. BOA provides * PHP-FPM 7.0, 5.6, 5.5, 5.4, 5.3 multi-install mode, configurable per site.  

We need to add a flag to the devshop installer for which versions of PHP you would like installed (besides the version of PHP that AEgir is running on) then provide a dropdown in the environment/project settings to allow you to set the version of PHP. Having a per environment setting might be handy for certain projects, not certain if this would cause any issues with the project's configuration, would need to ask Jon about that.

**Comment by jonpugh** [comment]
I don't want to support anything that isn't the default version for the OS. People can do that if they want to.
Let's start by supporting Ubuntu 16.04 which defaults to PHP7.

We can add it to the tests matrix.  Otherwise, I think this is out of scope until we get docker support really going.

**Comment by jonpugh** [comment]
> That said, maybe a better option would be to allow different versions of PHP to be used on the Aegir server as BOA Aegir does.

This would be great, but we aren't familiar enough with how it's done to support it. 

Perhaps @memtkmcc would be able to help get this feature more easily available in Aegir core?

**Comment by memtkmcc** [comment]
Running multiple PHP-FPM versions is the hardest thing -- it requires custom installation and config paths per version (which you need to standardize somehow), custom startup scripts, etc. so you just can't use standard OS specific PHP packages to make this happen.

We still build PHP from sources in BOA, which is kind of crazy, because it takes a lot of time during installation and version upgrades. It also needs custom maintenance and makes all extra PHP extensions installations and updates non-standard too.

That is why we plan to create our own packages at some point, to make it faster and less frustrating.

But once you have this part figured out and working, setting site specific PHP-FPM version is quite easy. The trick we use is based on managing little one-line config includes which automatically switch the path to the expected PHP-FPM socket (hence the need for standardization of naming convention in paths, also for sockets).

Of course you then need another script to manage these changes on the fly, and we have a bash script which reads the versions defined per site in a single plaintext config file the end user can manage easily.

So at the end it is all actually pretty complex, and we still discover edge case issues with these auto-updates.

I don't think it is possible to somehow include this trick in Aegir as-is now, at least perhaps until we or someone else could provide and maintain a custom built PHP packages to make it possible to use in the standard Aegir or DevShop installer.

That said, we could at least extract and publish the configuration switch trick code as a standalone Provision extension, perhaps?

**Comment by arttus** [comment]
"Let's start by supporting Ubuntu 16.04 which defaults to PHP7."

Jon has a valid point, maybe it really only makes sense to support multi-install mode on BOA. 

When and if BOA includes the Devshop front-end then we should add support for multi-install mode on BOA/Devshop. Until then duplicating functionality that BOA has already figured out is just reinventing the wheel.

Some documentation on how to add or change the existing version of PHP, might go a long way.

**Comment by jonpugh** [comment]
Thanks a ton for that detailed explanation, @memtkmcc . That's really helpful info.

I'm sold that this isn't worth it. We should put all of our efforts into [Hosting Docker](https://www.drupal.org/project/hosting_docker). Once that is done, changing PHP versions is done with a select list on node/%node/edit. That info is then written to a docker-compose.yml file, and the container is relaunched with the new version.

**Comment by jonpugh** [comment]
@arttus @kepford @madalinignisca DevShop has been fixed to work on Ubuntu16 which means it works well on PHP7 on any OS.

I'm going to recommend users use Ubuntu 16 for PHP7, or Ubuntu 14 for PHP5, and tell them if they want other combinations they can install different PHP versions manually (I'll find a good article on how to do that and include it in the documentation.)

**Comment by jonpugh** [comment]
I'm not content with forcing users to learn OS management to change PHP version.

@omega8cc has allowed this for years: https://www.drupal.org/project/provision/issues/1490160

We can simply add it to our VHOST templates and install the multi-php packages from ondrej
